src/main/scala/tech/sourced/api/provider/RepositoryProvider.scala

```
@@ -0,0 +1,91 @@
+package tech.sourced.api.provider
+
+import java.io.File
+import java.nio.file.Paths
+
+import org.apache.commons.io.FileUtils
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.spark.input.PortableDataStream
+import org.eclipse.jgit.lib.{Repository, RepositoryBuilder}
+import tech.sourced.api.util.MD5Gen
+import tech.sourced.siva.SivaReader
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+class RepositoryProvider(localPath: String) {
+  val repositories: mutable.Map[String, Repository] = mutable.Map()
+
+  def get(conf: Configuration, path: String): Repository = {
+    repositories.get(path) match {
+      case Some(r) =>
+        r.incrementOpen()
+        r
+      case None => RepositoryProvider.genRepository(conf, path, localPath)
+    }
+  }
+
+  def get(pds: PortableDataStream): Repository =
+    this.get(pds.getConfiguration, pds.getPath())
+
+  def close(path: String): Unit = {
+    repositories.get(path) match {
+      case Some(r) =>
+        r.close()
+      // TODO maybe others are using this repository instance
+      // FileUtils.deleteQuietly(r.getDirectory)
+      case None =>
+    }
+  }
+}
+
+object RepositoryProvider {
+  var provider: RepositoryProvider = _
+
+  def apply(localPath: String): RepositoryProvider = {
```

if this is a singleton, what about moving everything to the `RepositoryProvider` object instead of having the class and manually manage the singleton?

Also, if you do `RepositoryProvider("foo")` and then `RepositoryProvider("bar")` what you get is a repository provider with `"foo"` as localPath, which is a bit misleading.

---

Yep, you're right. I did this as it is because in a near future maybe we are going need a repository provider per path.

In the way that is now we should throw an error if a provider instance exists but not with the same `localPath` provided. What do you think?

---

Ok, makes sense. Yeah, maybe we should throw an error in that case :+1: